Fails to build on OpenSSL-4.x-dev

### Problem

OpenSSL-4.x officially removed support for Engine API from the code. And the current `libp11` code does not seem to allow to build a Provider without building the Engine first.

```text
.  .  .
libp11 has been configured with the following options:

Version:                 0.4.17_git
libp11 directory:        /Users/ur20980/openssl-4/lib
Engine directory:        /Users/ur20980/openssl-4/lib/engines-4/
Provider directory:      /Users/ur20980/openssl-3/lib/ossl-modules
Default PKCS11 module:   /opt/local/lib/p11-kit-proxy.dylib
API doc support:         no
Static PKCS#11 engine    no

Host:                    aarch64-apple-darwin25.2.0
Compiler:                clang
Preprocessor flags:      -g 
Compiler flags:          -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -pthread
Linker flags:            -g 
Libraries:               -lpthread 

OPENSSL_CFLAGS:          -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -g -I/Users/ur20980/openssl-4/include
OPENSSL_LIBS:            -L/Users/ur20980/openssl-4/lib -lssl -lcrypto
.  .  .

libtool: compile:  clang -DHAVE_CONFIG_H -I. -g -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -g -I/Users/ur20980/openssl-4/include -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -pthread -MT libp11_la-p11_ec.lo -MD -MP -MF .deps/libp11_la-p11_ec.Tpo -c p11_ec.c -o libp11_la-p11_ec.o >/dev/null 2>&1
p11_pkey.c:530:33: error: call to undeclared function 'EVP_PKEY_meth_find'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  530 |         orig_meth = (EVP_PKEY_METHOD *)EVP_PKEY_meth_find(EVP_PKEY_RSA);
      |                                        ^
p11_pkey.c:530:14: warning: cast to 'EVP_PKEY_METHOD *' (aka 'struct evp_pkey_method_st *') from smaller integer type 'int' [-Wint-to-pointer-cast]
  530 |         orig_meth = (EVP_PKEY_METHOD *)EVP_PKEY_meth_find(EVP_PKEY_RSA);
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p11_pkey.c:531:2: error: call to undeclared function 'EVP_PKEY_meth_get_sign'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  531 |         EVP_PKEY_meth_get_sign(orig_meth,
      |         ^
p11_pkey.c:533:2: error: call to undeclared function 'EVP_PKEY_meth_get_decrypt'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  533 |         EVP_PKEY_meth_get_decrypt(orig_meth,
      |         ^
p11_pkey.c:537:13: error: call to undeclared function 'EVP_PKEY_meth_new'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  537 |         new_meth = EVP_PKEY_meth_new(EVP_PKEY_RSA,
      |                    ^
p11_pkey.c:537:11: error: incompatible integer to pointer conversion assigning to 'EVP_PKEY_METHOD *' (aka 'struct evp_pkey_method_st *') from 'int' [-Wint-conversion]
  537 |         new_meth = EVP_PKEY_meth_new(EVP_PKEY_RSA,
      |                  ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  538 |                 EVP_PKEY_FLAG_AUTOARGLEN);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~
p11_pkey.c:540:2: error: call to undeclared function 'EVP_PKEY_meth_copy'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  540 |         EVP_PKEY_meth_copy(new_meth, orig_meth);
      |         ^
p11_pkey.c:542:2: error: call to undeclared function 'EVP_PKEY_meth_set_sign'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  542 |         EVP_PKEY_meth_set_sign(new_meth,
      |         ^
p11_pkey.c:544:2: error: call to undeclared function 'EVP_PKEY_meth_set_decrypt'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  544 |         EVP_PKEY_meth_set_decrypt(new_meth,
      |         ^
p11_pkey.c:674:33: error: call to undeclared function 'EVP_PKEY_meth_find'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  674 |         orig_meth = (EVP_PKEY_METHOD *)EVP_PKEY_meth_find(EVP_PKEY_EC);
      |                                        ^
p11_pkey.c:674:14: warning: cast to 'EVP_PKEY_METHOD *' (aka 'struct evp_pkey_method_st *') from smaller integer type 'int' [-Wint-to-pointer-cast]
  674 |         orig_meth = (EVP_PKEY_METHOD *)EVP_PKEY_meth_find(EVP_PKEY_EC);
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p11_pkey.c:675:2: error: call to undeclared function 'EVP_PKEY_meth_get_sign'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  675 |         EVP_PKEY_meth_get_sign(orig_meth,
      |         ^
p11_pkey.c:678:13: error: call to undeclared function 'EVP_PKEY_meth_new'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  678 |         new_meth = EVP_PKEY_meth_new(EVP_PKEY_EC, 0);
      |                    ^
p11_pkey.c:678:11: error: incompatible integer to pointer conversion assigning to 'EVP_PKEY_METHOD *' (aka 'struct evp_pkey_method_st *') from 'int' [-Wint-conversion]
  678 |         new_meth = EVP_PKEY_meth_new(EVP_PKEY_EC, 0);
      |                  ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p11_pkey.c:680:2: error: call to undeclared function 'EVP_PKEY_meth_copy'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  680 |         EVP_PKEY_meth_copy(new_meth, orig_meth);
      |         ^
p11_pkey.c:682:2: error: call to undeclared function 'EVP_PKEY_meth_set_sign'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  682 |         EVP_PKEY_meth_set_sign(new_meth,
      |         ^
2 warnings and 14 errors generated.
make[2]: *** [libp11_la-p11_pkey.lo] Error 1
$
$ openssl4 version
OpenSSL 4.0.0-dev  (Library: OpenSSL 4.0.0-dev )
$
```

### Suggested solution

Split the Provider code out cleanly, and make it possible to build Provider only via `make` target.

> libp11 compiles against the current OpenSSL 4.0.0-dev again

Sorry, not for me.

On OpenSSL-3.6.2 it compiles - but all the tests are marked 'SKIP". On OpenSSL-4.x is does not even compile - see below.

> [@mouse07410](https://github.com/mouse07410) Kindly try the new code and let us know your testing results. Support for paddings other than PKCS#1 had to be rewritten to use the provider API instead of the API removed in OpenSSL 4.0.

I haven't had any luck building `libp11` with OpenSSL-4 ever since they removed the deprecated API. That includes the current master. I don't want to make unnecessary noise and complicate this development. Here are my results:

Here's 
[conf-out.txt](https://github.com/user-attachments/files/26985936/conf-out.txt)

```text
.  .  .
libtool: link: clang -dynamiclib -Wl,-undefined -Wl,dynamic_lookup -Wl,-no_fixup_chains -o .libs/libp11.4.dylib  .libs/libp11_la-libpkcs11.o .libs/libp11_la-p11_attr.o .libs/libp11_la-p11_cert.o .libs/libp11_la-p11_ckr.o .libs/libp11_la-p11_key.o .libs/libp11_la-p11_load.o .libs/libp11_la-p11_misc.o .libs/libp11_la-p11_rsa.o .libs/libp11_la-p11_ec.o .libs/libp11_la-p11_eddsa.o .libs/libp11_la-p11_pkey.o .libs/libp11_la-p11_slot.o .libs/libp11_la-p11_front.o .libs/libp11_la-p11_atfork.o   -Wl,-force_load,./.libs/libp11_err.a  -L/Users/mouse07410/openssl-4/lib -lssl -lcrypto -lpthread  -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -g -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -g   -pthread -install_name  /Users/mouse07410/openssl-4/lib/libp11.4.dylib -compatibility_version 11 -current_version 11.4 -Wl,-exported_symbols_list,.libs/libp11-symbols.expsym
libtool: link: ( cd ".libs" && rm -f "libeng_err.la" && ln -s "../libeng_err.la" "libeng_err.la" )
/bin/sh ../libtool  --tag=CC   --mode=link clang   -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -g -I/Users/mouse07410/openssl-4/include -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -pthread  -module -shared -shrext .dylib -avoid-version -export-symbols "./pkcs11prov.exports" -g  -o pkcs11prov.la -rpath /Users/mouse07410/openssl-4/lib/ossl-modules pkcs11prov_la-provider.lo pkcs11prov_la-provider_helpers.lo pkcs11prov_la-util_uri.lo  libp11_err.la libp11_la-libpkcs11.lo libp11_la-p11_attr.lo libp11_la-p11_cert.lo libp11_la-p11_ckr.lo libp11_la-p11_key.lo libp11_la-p11_load.lo libp11_la-p11_misc.lo libp11_la-p11_rsa.lo libp11_la-p11_ec.lo libp11_la-p11_eddsa.lo libp11_la-p11_pkey.lo libp11_la-p11_slot.lo libp11_la-p11_front.lo libp11_la-p11_atfork.lo -L/Users/mouse07410/openssl-4/lib -lssl -lcrypto -lpthread 
Undefined symbols for architecture arm64:
  "_PKCS11_pkey_meths", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libp11.la] Error 1
make[2]: *** Waiting for unfinished jobs....
libtool: link: /opt/local/bin/gsed -e 's|^|_|' < ./pkcs11prov.exports > .libs/pkcs11prov-symbols.expsym
libtool: link: clang -Wl,-undefined -Wl,dynamic_lookup -Wl,-no_fixup_chains -o .libs/pkcs11prov.dylib -bundle  .libs/pkcs11prov_la-provider.o .libs/pkcs11prov_la-provider_helpers.o .libs/pkcs11prov_la-util_uri.o .libs/libp11_la-libpkcs11.o .libs/libp11_la-p11_attr.o .libs/libp11_la-p11_cert.o .libs/libp11_la-p11_ckr.o .libs/libp11_la-p11_key.o .libs/libp11_la-p11_load.o .libs/libp11_la-p11_misc.o .libs/libp11_la-p11_rsa.o .libs/libp11_la-p11_ec.o .libs/libp11_la-p11_eddsa.o .libs/libp11_la-p11_pkey.o .libs/libp11_la-p11_slot.o .libs/libp11_la-p11_front.o .libs/libp11_la-p11_atfork.o   -Wl,-force_load,./.libs/libp11_err.a  -L/Users/mouse07410/openssl-4/lib -lssl -lcrypto -lpthread  -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -g -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -g   -pthread -Wl,-exported_symbols_list,.libs/pkcs11prov-symbols.expsym
libtool: link: ( cd ".libs" && rm -f "pkcs11prov.la" && ln -s "../pkcs11prov.la" "pkcs11prov.la" )
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
+ OPENSSL_DIR=/Users/mouse07410/openssl-4
+ make check
+ tee check-out.txt
Making check in src
/bin/sh ../libtool  --tag=CC   --mode=link clang   -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -g -I/Users/mouse07410/openssl-4/include -O3 -std=gnu18 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include -pthread  -module -shared -shrext .dylib -avoid-version -export-symbols "./pkcs11.exports" -g  -o pkcs11.la -rpath /Users/mouse07410/openssl-4/lib pkcs11_la-eng_front.lo pkcs11_la-eng_back.lo pkcs11_la-util_uri.lo  libp11_err.la libeng_err.la libp11_la-libpkcs11.lo libp11_la-p11_attr.lo libp11_la-p11_cert.lo libp11_la-p11_ckr.lo libp11_la-p11_key.lo libp11_la-p11_load.lo libp11_la-p11_misc.lo libp11_la-p11_rsa.lo libp11_la-p11_ec.lo libp11_la-p11_eddsa.lo libp11_la-p11_pkey.lo libp11_la-p11_slot.lo libp11_la-p11_front.lo libp11_la-p11_atfork.lo -L/Users/mouse07410/openssl-4/lib -lssl -lcrypto -lpthread 
libtool: link: /opt/local/bin/gsed -e 's|^|_|' < ./pkcs11.exports > .libs/pkcs11-symbols.expsym
libtool: link: clang -Wl,-undefined -Wl,dynamic_lookup -Wl,-no_fixup_chains -o .libs/pkcs11.dylib -bundle  .libs/pkcs11_la-eng_front.o .libs/pkcs11_la-eng_back.o .libs/pkcs11_la-util_uri.o .libs/libp11_la-libpkcs11.o .libs/libp11_la-p11_attr.o .libs/libp11_la-p11_cert.o .libs/libp11_la-p11_ckr.o .libs/libp11_la-p11_key.o .libs/libp11_la-p11_load.o .libs/libp11_la-p11_misc.o .libs/libp11_la-p11_rsa.o .libs/libp11_la-p11_ec.o .libs/libp11_la-p11_eddsa.o .libs/libp11_la-p11_pkey.o .libs/libp11_la-p11_slot.o .libs/libp11_la-p11_front.o .libs/libp11_la-p11_atfork.o   -Wl,-force_load,./.libs/libp11_err.a -Wl,-force_load,./.libs/libeng_err.a  -L/Users/mouse07410/openssl-4/lib -lssl -lcrypto -lpthread  -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -g -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -g   -pthread -Wl,-exported_symbols_list,.libs/pkcs11-symbols.expsym
Undefined symbols for architecture arm64:
  "_bind_engine", referenced from:
      <initial-undefines>
  "_v_check", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

No, after fixing the bug in my script that you pointed out, still the same problem (Apple Silicon, MacOS Tahoe 26.4.1):

```
  CCLD     pkcs11prov.la
Undefined symbols for architecture arm64:
  "_PKCS11_pkey_meths", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libp11.la] Error 1
make[2]: *** Waiting for unfinished jobs....
Undefined symbols for architecture arm64:
  "_bind_engine", referenced from:
      <initial-undefines>
  "_v_check", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [pkcs11.la] Error 1
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
```

I **hard-coded `PKCS11_TOOL=/usr/local/bin/pkcs11-tool`** and re-run the tests:

```text
$ make check        
Making check in src
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-local
cd .libs && ln -s -f pkcs11.dylib libpkcs11.dylib
Making check in doc
make[1]: Nothing to be done for `check'.
Making check in examples
make[1]: Nothing to be done for `check'.
Making check in tests
/Applications/Xcode.app/Contents/Developer/usr/bin/make  openssl_version evp-sign evp-sign-prov fork-test fork-change-slot fork-change-slot-prov list-tokens rsa-pss-sign rsa-pss-sign-prov rsa-oaep rsa-oaep-prov check-privkey check-privkey-prov dup-key dup-key-prov check-all-prov rsa-keygen ec-keygen ed25519-keygen ed448-keygen ed25519-keygen-prov ed448-keygen-prov \
	  rsa-testpkcs11.softhsm rsa-testfork.softhsm rsa-testlistkeys.softhsm rsa-testlistkeys_ext.softhsm rsa-evp-sign.softhsm rsa-pss-sign.softhsm rsa-oaep.softhsm rsa-check-privkey.softhsm rsa-cert-store.softhsm rsa-keygen.softhsm ec-testfork.softhsm ec-evp-sign.softhsm ec-check-privkey.softhsm ec-cert-store.softhsm ec-copy.softhsm ec-keygen.softhsm ed25519-keygen.softhsm ed448-keygen.softhsm fork-change-slot.softhsm case-insensitive.softhsm pkcs11-uri-pin-source.softhsm pkcs11-uri-without-token.softhsm search-all-matching-tokens.softhsm provider-rsa-evp-sign.softhsm provider-rsa-pss-sign.softhsm provider-rsa-oaep.softhsm provider-rsa-check-privkey.softhsm provider-rsa-check-all.softhsm provider-ec-evp-sign.softhsm provider-ec-check-privkey.softhsm provider-ec-check-all.softhsm provider-ec-copy.softhsm provider-ed25519-check-privkey.softhsm provider-ed448-check-privkey.softhsm provider-ed25519-keygen.softhsm provider-ed448-keygen.softhsm provider-fork-change-slot.softhsm provider-case-insensitive.softhsm provider-pkcs11-uri-without-token.softhsm provider-search-all-matching-tokens.softhsm rsa-cert.der rsa-privkey.der rsa-pubkey.der ec-cert.der ec-privkey.der ec-pubkey.der ed25519-cert.der ed25519-privkey.der ed25519-pubkey.der ed448-cert.der ed448-privkey.der ed448-pubkey.der
make[2]: `openssl_version' is up to date.
make[2]: `evp-sign' is up to date.
make[2]: `evp-sign-prov' is up to date.
make[2]: `fork-test' is up to date.
make[2]: `fork-change-slot' is up to date.
make[2]: `fork-change-slot-prov' is up to date.
make[2]: `list-tokens' is up to date.
make[2]: `rsa-pss-sign' is up to date.
make[2]: `rsa-pss-sign-prov' is up to date.
make[2]: `rsa-oaep' is up to date.
make[2]: `rsa-oaep-prov' is up to date.
make[2]: `check-privkey' is up to date.
make[2]: `check-privkey-prov' is up to date.
make[2]: `dup-key' is up to date.
make[2]: `dup-key-prov' is up to date.
make[2]: `check-all-prov' is up to date.
make[2]: `rsa-keygen' is up to date.
make[2]: `ec-keygen' is up to date.
make[2]: `ed25519-keygen' is up to date.
make[2]: `ed448-keygen' is up to date.
make[2]: `ed25519-keygen-prov' is up to date.
make[2]: `ed448-keygen-prov' is up to date.
make[2]: Nothing to be done for `rsa-testpkcs11.softhsm'.
make[2]: Nothing to be done for `rsa-testfork.softhsm'.
make[2]: Nothing to be done for `rsa-testlistkeys.softhsm'.
make[2]: Nothing to be done for `rsa-testlistkeys_ext.softhsm'.
make[2]: Nothing to be done for `rsa-evp-sign.softhsm'.
make[2]: Nothing to be done for `rsa-pss-sign.softhsm'.
make[2]: Nothing to be done for `rsa-oaep.softhsm'.
make[2]: Nothing to be done for `rsa-check-privkey.softhsm'.
make[2]: Nothing to be done for `rsa-cert-store.softhsm'.
make[2]: Nothing to be done for `rsa-keygen.softhsm'.
make[2]: Nothing to be done for `ec-testfork.softhsm'.
make[2]: Nothing to be done for `ec-evp-sign.softhsm'.
make[2]: Nothing to be done for `ec-check-privkey.softhsm'.
make[2]: Nothing to be done for `ec-cert-store.softhsm'.
make[2]: Nothing to be done for `ec-copy.softhsm'.
make[2]: Nothing to be done for `ec-keygen.softhsm'.
make[2]: Nothing to be done for `ed25519-keygen.softhsm'.
make[2]: Nothing to be done for `ed448-keygen.softhsm'.
make[2]: Nothing to be done for `fork-change-slot.softhsm'.
make[2]: Nothing to be done for `case-insensitive.softhsm'.
make[2]: Nothing to be done for `pkcs11-uri-pin-source.softhsm'.
make[2]: Nothing to be done for `pkcs11-uri-without-token.softhsm'.
make[2]: Nothing to be done for `search-all-matching-tokens.softhsm'.
make[2]: Nothing to be done for `provider-rsa-evp-sign.softhsm'.
make[2]: Nothing to be done for `provider-rsa-pss-sign.softhsm'.
make[2]: Nothing to be done for `provider-rsa-oaep.softhsm'.
make[2]: Nothing to be done for `provider-rsa-check-privkey.softhsm'.
make[2]: Nothing to be done for `provider-rsa-check-all.softhsm'.
make[2]: Nothing to be done for `provider-ec-evp-sign.softhsm'.
make[2]: Nothing to be done for `provider-ec-check-privkey.softhsm'.
make[2]: Nothing to be done for `provider-ec-check-all.softhsm'.
make[2]: Nothing to be done for `provider-ec-copy.softhsm'.
make[2]: Nothing to be done for `provider-ed25519-check-privkey.softhsm'.
make[2]: Nothing to be done for `provider-ed448-check-privkey.softhsm'.
make[2]: Nothing to be done for `provider-ed25519-keygen.softhsm'.
make[2]: Nothing to be done for `provider-ed448-keygen.softhsm'.
make[2]: Nothing to be done for `provider-fork-change-slot.softhsm'.
make[2]: Nothing to be done for `provider-case-insensitive.softhsm'.
make[2]: Nothing to be done for `provider-pkcs11-uri-without-token.softhsm'.
make[2]: Nothing to be done for `provider-search-all-matching-tokens.softhsm'.
make[2]: Nothing to be done for `rsa-cert.der'.
make[2]: Nothing to be done for `rsa-privkey.der'.
make[2]: Nothing to be done for `rsa-pubkey.der'.
make[2]: Nothing to be done for `ec-cert.der'.
make[2]: Nothing to be done for `ec-privkey.der'.
make[2]: Nothing to be done for `ec-pubkey.der'.
make[2]: Nothing to be done for `ed25519-cert.der'.
make[2]: Nothing to be done for `ed25519-privkey.der'.
make[2]: Nothing to be done for `ed25519-pubkey.der'.
make[2]: Nothing to be done for `ed448-cert.der'.
make[2]: Nothing to be done for `ed448-privkey.der'.
make[2]: Nothing to be done for `ed448-pubkey.der'.
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: rsa-testpkcs11.softhsm
FAIL: rsa-testfork.softhsm
PASS: rsa-testlistkeys.softhsm
PASS: rsa-testlistkeys_ext.softhsm
SKIP: rsa-evp-sign.softhsm
SKIP: rsa-pss-sign.softhsm
SKIP: rsa-oaep.softhsm
SKIP: rsa-check-privkey.softhsm
PASS: rsa-cert-store.softhsm
SKIP: rsa-keygen.softhsm
FAIL: ec-testfork.softhsm
SKIP: ec-evp-sign.softhsm
SKIP: ec-check-privkey.softhsm
PASS: ec-cert-store.softhsm
SKIP: ec-copy.softhsm
SKIP: ec-keygen.softhsm
SKIP: ed25519-keygen.softhsm
SKIP: ed448-keygen.softhsm
SKIP: fork-change-slot.softhsm
SKIP: case-insensitive.softhsm
SKIP: pkcs11-uri-pin-source.softhsm
SKIP: pkcs11-uri-without-token.softhsm
SKIP: search-all-matching-tokens.softhsm
PASS: provider-rsa-evp-sign.softhsm
PASS: provider-rsa-pss-sign.softhsm
PASS: provider-rsa-oaep.softhsm
PASS: provider-rsa-check-privkey.softhsm
PASS: provider-rsa-check-all.softhsm
PASS: provider-ec-evp-sign.softhsm
PASS: provider-ec-check-privkey.softhsm
PASS: provider-ec-check-all.softhsm
PASS: provider-ec-copy.softhsm
PASS: provider-ed25519-check-privkey.softhsm
PASS: provider-ed448-check-privkey.softhsm
PASS: provider-ed25519-keygen.softhsm
PASS: provider-ed448-keygen.softhsm
FAIL: provider-fork-change-slot.softhsm
PASS: provider-case-insensitive.softhsm
PASS: provider-pkcs11-uri-without-token.softhsm
FAIL: provider-search-all-matching-tokens.softhsm
============================================================================
Testsuite summary for libp11 0.4.19_git
============================================================================
# TOTAL: 40
# PASS:  20
# SKIP:  16
# XFAIL: 0
# FAIL:  4
# XPASS: 0
# ERROR: 0
============================================================================
See tests/test-suite.log for debugging.
============================================================================
make[3]: *** [test-suite.log] Error 1
make[2]: *** [check-TESTS] Error 2
make[1]: *** [check-am] Error 2
make: *** [check-recursive] Error 1
```

From the log:
```text
FAIL: provider-fork-change-slot.softhsm
=======================================

Current directory: /Users/mouse07410/src/libp11/tests
Source directory: .
Output directory: output.76902
SoftHSM library found: /opt/local/lib/softhsm/libsofthsm2.so
pkcs11-tool found: /usr/local/bin/pkcs11-tool
Compiled with: OpenSSL 4.1.0-dev  (Library: OpenSSL 4.1.0-dev )
Detected system: darwin25
OPENSSL_ENGINES=../src/.libs/
SoftHSM tool: /opt/local/bin/softhsm2-util
Configuration: output.76902/softhsm-testpkcs11.config
***************************************
* Initializing token token1 ... ok
***************************************
* Initializing token token2 ... ok
* Generating an RSA key pair on the token token2
Key pair generated:
Private Key Object; RSA  2048 bits
  Modulus:    d9d20f4856babe48b381e98fecfc164c11fc05e6bab24f7003b1c3d78a4f6afc
              2c81f2a804160835edf4a293809d78f3e1984b81479eed8c73ca9d300b3dc88b
              340c3dbbfe3e080e3918d73a6a4b5cdc11ba632056e943e1f6177d36ace34f1c
              46c929a74c0a82d6388cc7582815c2120bda2c761996e26ccf2c3e7eed932b2a
              4cc444805444ccf33d5d8ccb7e26cd87d3e13a0d0230248a4fa8ee46d1000a6d
              21e90943f27663385356016df736d7881609e60bc78af2fc3ebef7d69e56d502
              c02840e7de5cab8a265d91b1a7c2c11f6a50431b7d164578edd4c4f6c16661a8
              f4ec5a62d7e55ae0dd9edaf396fa16e6e9f28e60357ef1ccd0e06ee3207ddfa3
  Public exp: 65537 (0x010001)
  label:      pkey
  ID:         16909060 (0x01:02:03:04)
  Usage:      decrypt, sign, signRecover, unwrap
  Access:     sensitive, always sensitive, never extractable, local
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=1d44c64c69b53eef;token=token2;id=%01%02%03%04;object=pkey;type=private
Public Key Object; RSA  2048 bits
  Modulus:    d9d20f4856babe48b381e98fecfc164c11fc05e6bab24f7003b1c3d78a4f6afc
              2c81f2a804160835edf4a293809d78f3e1984b81479eed8c73ca9d300b3dc88b
              340c3dbbfe3e080e3918d73a6a4b5cdc11ba632056e943e1f6177d36ace34f1c
              46c929a74c0a82d6388cc7582815c2120bda2c761996e26ccf2c3e7eed932b2a
              4cc444805444ccf33d5d8ccb7e26cd87d3e13a0d0230248a4fa8ee46d1000a6d
              21e90943f27663385356016df736d7881609e60bc78af2fc3ebef7d69e56d502
              c02840e7de5cab8a265d91b1a7c2c11f6a50431b7d164578edd4c4f6c16661a8
              f4ec5a62d7e55ae0dd9edaf396fa16e6e9f28e60357ef1ccd0e06ee3207ddfa3
  Public exp: 65537 (0x010001)
  label:      pkey
  ID:         16909060 (0x01:02:03:04)
  Usage:      encrypt, verify, verifyRecover, wrap
  Access:     local
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=1d44c64c69b53eef;token=token2;id=%01%02%03%04;object=pkey;type=public
OPENSSL_MODULES=../src/.libs/
PKCS11_MODULE_PATH=/opt/local/lib/softhsm/libsofthsm2.so
Compiled with: OpenSSL 4.1.0-dev  (Library: OpenSSL 4.1.0-dev )
No custom UI method provided, using the default UI method.
Child 76968 terminated by signal #4
pid 76953 is the parent
Provider "pkcs11prov" set.
Provider "default" set.
PKCS#11: Initializing the module: /opt/local/lib/softhsm/libsofthsm2.so
- [1358744719] SoftHSM slot ID 0x50fccc8  login                                 (token1)
- [1773485807] SoftHSM slot ID 0x69b53ee  login                                 (token2)
- [2] SoftHSM slot ID 0x2        uninitialized, login                  (no label)
Searching slots without login for an initialized token containing private key  label=pkey
Found slot: SoftHSM slot ID 0x69b53eef
Found initialized token: token2
Searching slots with login for an initialized token containing private key  label=pkey
Found slot: SoftHSM slot ID 0x69b53eef
Found initialized token: token2
Found 1 private key:
   1 P  id=01020304 label=pkey
Returning last matching private key: id=01020304 label=pkey
Private key found.
spawned 76968
* Removing token token1
Found token (dc047175-1b28-2687-9397-d12bd0fccc8f) with matching token label.
The token (output.76902/softhsm-testpkcs11.db/dc047175-1b28-2687-9397-d12bd0fccc8f) has been deleted.
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
Fork test failed.
FAIL provider-fork-change-slot.softhsm (exit status: 1)

SKIP: rsa-oaep.softhsm
======================

Current directory: /Users/mouse07410/src/libp11/tests
Source directory: .
Output directory: output.74437
SoftHSM library found: /opt/local/lib/softhsm/libsofthsm2.so
pkcs11-tool found: /usr/local/bin/pkcs11-tool
Compiled with: OpenSSL 4.1.0-dev  (Library: OpenSSL 4.1.0-dev )
Detected system: darwin25
OPENSSL_ENGINES=../src/.libs/
SoftHSM tool: /opt/local/bin/softhsm2-util
Configuration: output.74437/softhsm-testpkcs11.config
***************************************
* Initializing token libp11-0 ... ok
* Importing the rsa privkey object id=01020304 into the token libp11-0 ... ok
* Importing the rsa pubkey object id=01020304 into the token libp11-0 ... ok
* Importing the rsa cert object id=01020304 into the token libp11-0 ... ok
***************************************
* Listing objects on the token libp11-0
***************************************
Certificate Object; type = X.509 cert
  label:      server-key-0
  subject:    DN: CN=server-1
  serial:     5587F1771D95BBCEC82B19F9
  ID:         16909060 (0x01:02:03:04)
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=e8511dee8f30518e;token=libp11-0;id=%01%02%03%04;object=server-key-0;type=cert
Private Key Object; RSA  2048 bits
  Modulus:    b2c137838b8eb3f5d922b701835bf6c493b1e6d39de839e0d1bf3155ab2c41b2
              1586e20e9ae6f840b2defbc39c5508ba41038cb92cd4aee8e4c89864149b2e1b
              51de6b67d3baa587b3d3e387d11fa7a72a2f21c50953a20ee1feb00f913c8bbc
              7efedfb3a11f1d3e67d5806115de421168fd1413529740509294964b4fe5c926
              041f689df9eeacc7eaa207f4852040d5f826cfa390830552323f2c4e8de6abbd
              1f5405869267aaccfebccda36fe0468328fa55db8895320e1c59ccd31e85f946
              46bfb6a780752011cc1a586af2dfceb5ee502f863131d45901304a3db3283bcb
              c3a2077a853771c4396db040d6b943b42a40bc12f7c736f5fa81160820bcd5ff
  Public exp: 65537 (0x010001)
  label:      server-key-0
  ID:         16909060 (0x01:02:03:04)
  Usage:      decrypt, sign, signRecover, unwrap
  Access:     sensitive
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=e8511dee8f30518e;token=libp11-0;id=%01%02%03%04;object=server-key-0;type=private
Public Key Object; RSA  2048 bits
  Modulus:    b2c137838b8eb3f5d922b701835bf6c493b1e6d39de839e0d1bf3155ab2c41b2
              1586e20e9ae6f840b2defbc39c5508ba41038cb92cd4aee8e4c89864149b2e1b
              51de6b67d3baa587b3d3e387d11fa7a72a2f21c50953a20ee1feb00f913c8bbc
              7efedfb3a11f1d3e67d5806115de421168fd1413529740509294964b4fe5c926
              041f689df9eeacc7eaa207f4852040d5f826cfa390830552323f2c4e8de6abbd
              1f5405869267aaccfebccda36fe0468328fa55db8895320e1c59ccd31e85f946
              46bfb6a780752011cc1a586af2dfceb5ee502f863131d45901304a3db3283bcb
              c3a2077a853771c4396db040d6b943b42a40bc12f7c736f5fa81160820bcd5ff
  Public exp: 65537 (0x010001)
  label:      server-key-0
  ID:         16909060 (0x01:02:03:04)
  Usage:      encrypt, verify, verifyRecover, wrap
  Access:     none
  uri:        pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=e8511dee8f30518e;token=libp11-0;id=%01%02%03%04;object=server-key-0;type=public
***************************************
Compiled with: OpenSSL 4.1.0-dev  (Library: OpenSSL 4.1.0-dev )
Skipped: ENGINE support not available
RSA-OAEP encryption/decryption test skipped.
SKIP rsa-oaep.softhsm (exit status: 77)
```

Apparently, (a) detection/location of `pkcs11-tool` is broken in this version, and (b) some tests still rely on the ENGINE API being present. (Not sure what that "kill" complaint was about.)

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -37,7 +37,7 @@ jobs:
             openssl: 'openssl@4'
             triplet: x64-osx
             compiler: clang
-            os: macOS-latest
+            os: macos-26
             generator: Unix Makefiles
           - id: macOS-3
             openssl: 'openssl@3'

--- a/.github/workflows/ci.yml.in
+++ b/.github/workflows/ci.yml.in
@@ -37,7 +37,7 @@ jobs:
             openssl: 'openssl@4'
             triplet: x64-osx
             compiler: clang
-            os: macOS-latest
+            os: macos-26
             generator: Unix Makefiles
           - id: macOS-3
             openssl: 'openssl@3'

--- a/configure.ac
+++ b/configure.ac
@@ -35,10 +35,14 @@ opensslversion="$( \
 	( $PKG_CONFIG --modversion --silence-errors libcrypto || \
     $PKG_CONFIG --modversion openssl ) | \
     sed 's/^\([0-9.]*\).*/\1/' )"
+
+LIBP11_BUILD_ENGINE="yes"
+
 case "$opensslversion" in
 	4.*) # Provider for OpenSSL 4.x
 	    LIBP11_LT_OLDEST="4"
-	    LIBP11_OSSL_PROVIDER="yes";;
+	    LIBP11_OSSL_PROVIDER="yes"
+	    LIBP11_BUILD_ENGINE="no";;
 	3.*) # Engines directory prefix for OpenSSL 3.x
 	    LIBP11_LT_OLDEST="3"
 	    LIBP11_OSSL_PROVIDER="yes"
@@ -55,12 +59,17 @@ case "$opensslversion" in
 esac
 
 case "$OSSL_PKG_VERSION" in
+	4.*)
+		AC_MSG_NOTICE([OpenSSL 4.x detected])
+		LIBP11_BUILD_ENGINE="no"
+		LIBP11_OSSL_PROVIDER="yes";;
 	3.*)
-		AC_MSG_NOTICE([3.*])
+		AC_MSG_NOTICE([OpenSSL 3.x detected])
 		LIBP11_OSSL_PROVIDER="yes";;
 esac
 
 AM_CONDITIONAL([LIBP11_OSSL_PROVIDER], [test x$LIBP11_OSSL_PROVIDER = xyes])
+AM_CONDITIONAL([LIBP11_BUILD_ENGINE], [test x$LIBP11_BUILD_ENGINE = xyes])
 
 # LT Version numbers, remember to change them just *before* a release.
 #   (Code changed:                      REVISION++)

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -11,14 +11,18 @@ EXTRA_DIST = Makefile.mak libp11.rc.in pkcs11.rc.in
 noinst_HEADERS= libp11-int.h pkcs11.h p11_pthread.h provider_helpers.h util.h
 include_HEADERS= libp11.h p11_err.h
 
+lib_LTLIBRARIES = libp11.la
+enginesexec_LTLIBRARIES =
+
+if LIBP11_BUILD_ENGINE
+enginesexec_LTLIBRARIES += pkcs11.la
+
 # Libraries to build (static-engine optional)
 if ENABLE_STATIC_ENGINE
-lib_LTLIBRARIES = libp11.la libpkcs11.la
-else
-lib_LTLIBRARIES = libp11.la
+lib_LTLIBRARIES += libpkcs11.la
+endif
 endif
 
-enginesexec_LTLIBRARIES = pkcs11.la
 pkgconfig_DATA = libp11.pc
 
 if LIBP11_OSSL_PROVIDER
@@ -36,7 +40,11 @@ SHARED_EXT=@SHARED_EXT@
 # specific source files without affecting the rest of the project.
 # ----------------------------------------------------------
 # Define all non-installed helper libraries in one assignment
-noinst_LTLIBRARIES = libeng_err.la libp11_err.la
+noinst_LTLIBRARIES = libp11_err.la
+
+if LIBP11_BUILD_ENGINE
+noinst_LTLIBRARIES += libeng_err.la
+endif
 
 # Helper library for p11_err.c
 libp11_err_la_SOURCES = p11_err.c
@@ -145,10 +153,15 @@ endif
 
 # OpenSSL older than 1.1.0 expected libpkcs11.so instead of pkcs11.so
 check-local: $(LTLIBRARIES)
+if LIBP11_BUILD_ENGINE
 	cd .libs && $(LN_S) -f pkcs11$(SHARED_EXT) libpkcs11$(SHARED_EXT)
+endif
 
 install-exec-hook:
+if LIBP11_BUILD_ENGINE
 	cd '$(DESTDIR)$(enginesexecdir)' && $(LN_S) -f pkcs11$(SHARED_EXT) libpkcs11$(SHARED_EXT)
+endif
+
 if LIBP11_OSSL_PROVIDER
 	cd '$(DESTDIR)$(providersexecdir)' && $(LN_S) -f pkcs11prov$(SHARED_EXT) libpkcs11$(SHARED_EXT)
 endif

--- a/src/libp11.h
+++ b/src/libp11.h
@@ -461,8 +461,13 @@ void *PKCS11_get_ec_key_method(void);
 ECDSA_METHOD *PKCS11_get_ecdsa_method(void);
 ECDH_METHOD *PKCS11_get_ecdh_method(void);
 #endif
+
+#if OPENSSL_VERSION_NUMBER < 0x40000000L
 int PKCS11_pkey_meths(ENGINE *e, EVP_PKEY_METHOD **pmeth,
 		const int **nids, int nid);
+#else /* OPENSSL_VERSION_NUMBER < 0x40000000L */
+int PKCS11_pkey_meths(void *e, void **pmeth, const int **nids, int nid);
+#endif /* OPENSSL_VERSION_NUMBER < 0x40000000L */
 
 /**
  * Load PKCS11 error strings

--- a/src/p11_pkey.c
+++ b/src/p11_pkey.c
@@ -997,6 +997,18 @@ int PKCS11_pkey_meths(ENGINE *e, EVP_PKEY_METHOD **pmeth,
 	return 0;
 }
 
+#else /* OPENSSL_VERSION_NUMBER < 0x40000000L */
+
+int PKCS11_pkey_meths(void *e, void **pmeth, const int **nids, int nid)
+{
+	(void)e;
+	(void)pmeth;
+	(void)nids;
+	(void)nid;
+	fprintf(stderr, "PKCS11_pkey_meths is not available: ENGINE support was disabled for OpenSSL 4.x\n");
+	return 0;
+}
+
 #endif /* OPENSSL_VERSION_NUMBER < 0x40000000L */
 
 /* vim: set noexpandtab: */

--- a/tests/common.sh
+++ b/tests/common.sh
@@ -38,11 +38,12 @@ SOFTHSM_SEARCH_PATHS=(
 	"/usr/lib"
 )
 
+# # Prefer bin dirs over Cellar; Cellar may contain multiple/old versions and slow down search
 PKCS11_TOOL_SEARCH_PATHS=(
-	"/opt/homebrew/Cellar"
 	"/opt/homebrew/bin"
 	"/usr/local/bin"
 	"/usr/bin"
+	"/opt/homebrew/Cellar"
 )
 
 # Locate the SoftHSM library
@@ -58,8 +59,12 @@ else
 fi
 
 # Locate the pkcs11-tool
-PKCS11_TOOL=$(find "${PKCS11_TOOL_SEARCH_PATHS[@]}" -type f -name "pkcs11-tool" \
-	-print -quit 2>/dev/null)
+PKCS11_TOOL=$(command -v pkcs11-tool 2>/dev/null || true)
+
+if [[ -z "${PKCS11_TOOL}" ]]; then
+	PKCS11_TOOL=$(find -L "${PKCS11_TOOL_SEARCH_PATHS[@]}" -type f \
+	-name "pkcs11-tool" 2>/dev/null | head -n 1)
+fi
 
 # Output the result
 if [[ -n "${PKCS11_TOOL}" ]]; then

